### PR TITLE
Update convert_ismapper_to_vcf.py

### PR DESCRIPTION
**Pull request by vrennie**
### FROM https://github.com/TORCH-Consortium/MAGMA/pull/215#issuecomment-2285745436

We will have a decision by Thursday morning.

For **short-term**:

- Remove the is-mapper 
- Need to keep SV analysis in the pipeline

For **long-term**: 

- Evaluate `gatk-sv` https://github.com/broadinstitute/gatk-sv?tab=readme-ov-file#annotate-vcf and https://gatk.broadinstitute.org/hc/en-us/articles/9022476791323-Structural-Variants 

For **mid-term**:

- Implement this temporary script based solution

image

### CODE REVIEW

**Comment by vrennie** [comment]
Hi @abhi18av yes indeed! The issue is that TBProfiler is not reporting these correctly. I need to debug with @mdediegofuertes tomorrow.

**Comment by abhi18av** [comment]
No worries, this might also reflect on the `tbdb/MAGMA` database we constructed so just keep me posted on this please :)

**Comment by abhi18av** [comment]
We will have a decision by Thursday morning.

For **short-term**:

    - Remove the is-mapper 
    - Need to keep SV analysis in the pipeline

For **long-term**: 

    - Evaluate `gatk-sv` https://github.com/broadinstitute/gatk-sv?tab=readme-ov-file#annotate-vcf

For **mid-term**:

    - Implement this temporary script based solution

image

**Comment by Wilma-chan** [comment]
Hi all,
I wrote the script that integrates the delly vcf, ISmapper vcf and TB profiler vcf, I attached here so abhi18av can  check it and integrated it.
[script_03.09.24.py.txt](https://github.com/user-attachments/files/16848637/script_03.09.24.py.txt)




import pysam
import pandas as pd
import json
import subprocess  # Import subprocess module

# Define file paths
delly_vcf = '/mnt/c/Users/peppe/OneDrive/Documentos/Escritorio/downloads/int/pipe2/IS6110.S011.filtered.delly.bcf'
ismapper_vcf = '/mnt/c/Users/peppe/OneDrive/Documentos/Escritorio/downloads/int/IS6110.S011.L1.A1.1.1.1.ismapper.vcf'
bed_file = '/mnt/c/Users/peppe/OneDrive/Documentos/Escritorio/downloads/int/pipe2/who_plus.bed'
existing_json_file = '/mnt/c/Users/peppe/OneDrive/Documentos/Escritorio/downloads/int/pipe2/IS6110.S011.results-4_TB_profiler.json'
updated_json_file = '/mnt/c/Users/peppe/OneDrive/Documentos/Escritorio/downloads/int/pipe2/IS6110.S011.results-4_TBprofiler_updated_fromdelly.json'
cleaned_json_file = '/mnt/c/Users/peppe/OneDrive/Documentos/Escritorio/downloads/int/pipe2/IS6110.S011.results-4_TBprofiler_cleaned_fromdelly.json'

# Output concatenated VCF file path
output_vcf = '/mnt/c/Users/peppe/OneDrive/Documentos/Escritorio/downloads/int/concatenated_output.vcf'

# Command to concatenate BCF/VCF files using bcftools
bcftools_command = [
    'bcftools', 'concat', '-o', output_vcf, '-O', 'v', delly_vcf, ismapper_vcf  # Corrected variable name
]

# Run the command
subprocess.run(bcftools_command, check=True)

print(f'Concatenated VCF file created at: {output_vcf}')


# Step 1: Load BED file into a DataFrame
bed_df = pd.read_csv(bed_file, sep='\t', header=None, names=['Chromosome', 'start', 'end', 'gene_code', 'gene_name', 'drug'])

# Create a dictionary to store BED intervals by chromosome
bed_intervals = {}
for _, row in bed_df.iterrows():
    chrom = row['Chromosome']
    if chrom not in bed_intervals:
        bed_intervals[chrom] = []
    bed_intervals[chrom].append(row)

# Step 2: Process the VCF file and create JSON output
vcf_in = pysam.VariantFile(output_vcf)
json_output = []

for record in vcf_in:
    vcf_chrom = record.chrom
    vcf_start = record.start
    vcf_end = record.stop  # Use record.stop to get the end position

    if vcf_chrom in bed_intervals:
        for bed_entry in bed_intervals[vcf_chrom]:
            bed_start = bed_entry['start']
            bed_end = bed_entry['end']

            # Check for overlap between VCF and BED intervals
            overlap = (
                (vcf_start <= bed_end and vcf_end >= bed_start) or
                (bed_start <= vcf_end and bed_end >= vcf_start)
            )
            
            if overlap:
                sv_type = record.info.get('SVTYPE', None)
                
                # Compute sv_len
                if sv_type in ['DEL', 'DUP', 'INV']:
                    vcf_svlen = vcf_end - vcf_start + 1
                elif sv_type == 'INS':
                    vcf_svlen = abs(record.info.get('SVLEN', record.info.get('INSLEN', 0)))
                else:
                    vcf_svlen = None

                # Determine change format
                change_format = sv_type.lower() if sv_type in ['INS', 'DEL', 'DUP', 'INV'] else 'unknown'

                # Create JSON record
                json_record = {
                    "chrom": vcf_chrom,
                    "pos": vcf_start + 1,  # VCF is 0-based, JSON is 1-based
                    "ref": record.ref,
                    "alt": record.alts[0],
                    "depth": None,
                    "freq": None,
                    "sv": True,
                    "filter": list(record.filter.keys())[0] if record.filter else "PASS",
                    "forward_reads": None,
                    "reverse_reads": None,
                    "sv_len": vcf_svlen,
                    "gene_id": bed_entry['gene_code'],
                    "gene_name": bed_entry['gene_name'],
                    "feature_id": None,
                    "type": None,
                    "change": f"c.{vcf_start + 1}_{vcf_end}{change_format}",
                    "nucleotide_change": f"c.{vcf_start + 1}_{vcf_end}{change_format}",
                    "protein_change": None,
                    "annotation": [
                        {
                            "type": "drug_resistance",
                            "drug": bed_entry['drug'],
                            "original_mutation": "LoF",
                            "confidence": " ",
                            "source": "",
                            "comment": ""
                        }
                    ],
                    "consequences": [
                        {
                            "gene_id": bed_entry['gene_code'],
                            "gene_name": bed_entry['gene_name'],
                            "feature_id": None,
                            "type": None,
                            "nucleotide_change": f"c.{vcf_start + 1}_{vcf_end}{change_format}",
                            "protein_change": None,
                            "annotation": [
                                {
                                    "type": "drug_resistance",
                                    "drug": bed_entry['drug'],
                                    "original_mutation": "LoF",
                                    "confidence": "",
                                    "source": "",
                                    "comment": ""
                                }
                            ]
                        }
                    ],
                    "drugs": [
                        {
                            "type": "drug_resistance",
                            "drug": bed_entry['drug'],
                            "original_mutation": "LoF",
                            "confidence": "",
                            "source": "",
                            "comment": ""
                        }
                    ],
                    "locus_tag": bed_entry['gene_code'],
                    "gene_associated_drugs": [bed_entry['drug']] if bed_entry['drug'] else []
                }

                # Append the record to the output list
                json_output.append(json_record)
                break

# Step 3: Update the existing JSON with new data
try:
    with open(existing_json_file, 'r') as file:
        existing_data = json.load(file)
except IOError as e:
    print(f"Error reading file {existing_json_file}: {e}")
    existing_data = {}

# Mark records from TBprofiler as "source: TBprofiler"
for record in existing_data.get('dr_variants', []):
    record['source'] = 'TBprofiler'

# Mark new records from Delly/Ismapper as "source: Delly/Ismapper"
for record in json_output:
    record['source'] = 'Delly/Ismapper'

# Combine both sets of records
combined_data = existing_data.get('dr_variants', []) + json_output

# Step 4: Remove duplicates, prioritizing TBprofiler records
unique_records = {}
for record in combined_data:
    key = (record["chrom"], record["pos"], record["alt"], record["sv_len"], record["gene_id"])
    
    # If the key exists and the existing record is from TBprofiler, skip adding the new one
    if key in unique_records:
        if unique_records[key]['source'] == 'TBprofiler':
            continue
    
    unique_records[key] = record

# Update with unique records, removing the 'source' key
existing_data["dr_variants"] = [v for k, v in unique_records.items()]

# Step 5: Save the cleaned data to a new JSON file
try:
    with open(cleaned_json_file, 'w') as file:
        json.dump(existing_data, file, indent=4)
    print(f"Cleaned JSON file saved: {cleaned_json_file}")
except IOError as e:
    print(f"Error writing file {cleaned_json_file}: {e}")

**Comment by vrennie** [comment]
@abhi18av could take a look at this? 

Thanks

**Comment by vrennie** [comment]
Hi @abhi18av can you please take a look at this this week?

**Comment by abhi18av** [comment]
Hi @vrennie  and @Wilma-chan , sorry for the delay.

Actually, I didn't take this forward due to some doubts. 

As per our last meeting, I thought we wanted to evaluate a different (possibly less hacky) approach to address this problem no? I am awaiting feedback on the thread titled "GATK SVAnnotate Testing"


Please let me know if the plan has changed and you'd like to iterate on this solution anyhow.

**Comment by vrennie** [comment]
Hi @abhi18av, we looked into GATK SV Annotate but it doesn't work well, giving many spurious results. We will proceed with this and once SNPEff fixes large insertions/deletions we can get rid of Wilma's patch

**Comment by abhi18av** [comment]
Oki doki, makes sense. I'll ping here once the script is integrated 👍

**Comment by Wilma-chan** [comment]
[heart]         Wilma Quispe Rojas reacted to your message:
________________________________

**Comment by abhi18av** [comment]
Quick update regarding the current status,

@Wilma-chan , I have incorporated your script in this branch after a small refactor to generalize it https://github.com/TORCH-Consortium/MAGMA/pull/215/commits/0ec0e49c9095e495e5d0f855b6855d0bc9ff96e9 


I'm now testing the pipeline with some relevant samples

**Comment by Wilma-chan** [comment]
[heart]         Wilma Quispe Rojas reacted to your message:
________________________________

**Comment by abhi18av** [comment]
Hi @vrennie , 

Before signing off for the weekend, I was trying to debug the `STRUCTURAL_VARIANTS_ANALYSIS_WF:BCFTOOLS_MERGE__DELLY` process which isn't generating the expected `joint.delly.vcf` and therefore not triggering Wilma's script.

I am working on patching that script itself so that we are able to incorporate it.

Though, as a word of caution, please remember this will add another layer of hard-debugging experience once this is done.

**Comment by Wilma-chan** [comment]
[sad]   Wilma Quispe Rojas reacted to your message:
________________________________

**Comment by vrennie** [comment]
Hi @abhi18av,

Thanks for the update on this. I was on leave all of last week. Did the patch work?

**Comment by abhi18av** [comment]
Hi Vincent,

Haven't finalized it yet, I took a break over the long weekend here :)

**Comment by abhi18av** [comment]
@vrennie , so the problem is with the bioinformatics logic in `STRUCTURAL_VARIANTS_ANALYSIS_WF:BCFTOOLS_MERGE__DELLY` which isn't able to merge the `VCF` and `BCF` files corresponding to the sample

### Attempt with `*gz` files 

image


### Attempt with uncompressed files 

```
abhinav@hemera ~/p/M/w/0/09f1757af057f9556c0a86bce700e5> bcftools concat NICD.BF01114045-S1.filtered.delly.bcf NICD.BF01114045-S1.L1.A1.1.1.1.i
smapper.vcf
Different sample names in NICD.BF01114045-S1.L1.A1.1.1.1.ismapper.vcf. Perhaps "bcftools merge" is what you are looking for?
```

**Comment by abhi18av** [comment]
@vrennie , a fix for this issue might need to be tweak a stage before the samples reach  `STRUCTURAL_VARIANTS_ANALYSIS_WF:BCFTOOLS_MERGE__DELLY` 🤔 


I must say that to add this to the pipeline, we must resolve this issue within a proper conda environment and samples. 

Happy to jump on a call to address this better, ideally Friday or next Wednesday.

**Comment by vrennie** [comment]
HI @TimHHH which run is that coming from? I was looking at this run but that doesn't include those samples:

https://cloud.seqera.io/orgs/TORCH/workspaces/XBS-Nextflow/watch/2Xy5nIrPsq1wyU

**Comment by TimHHH** [comment]
@vrennie I believe it is from Abhi's own test collection, he mentioned these samples above.

**Comment by abhi18av** [comment]
Hi @TimHHH , yes that's the case.

I am attaching the VCF files I used earlier in the test.

[test_data.zip](https://github.com/user-attachments/files/17150576/test_data.zip)

**Comment by abhi18av** [comment]
> HI @TimHHH which run is that coming from? I was looking at this run but that doesn't include those samples:
> 
> https://cloud.seqera.io/orgs/TORCH/workspaces/XBS-Nextflow/watch/2Xy5nIrPsq1wyU

Yeah, generally the dev stuff I monitor on a different workspace not to clog up the TORCH workspace.

**Comment by abhi18av** [comment]
Meeeting notes (27-SEPT-2024)
Attendees: @Wilma-chan @vrennie @abhi18av 

1. The problem seems to be with the assumptions regarding sample names in the `convert..ismapper.py` script 

2. Ideally we restrict the sample names which the pipeline accepts to avoid these kinds of surprises 

3. NICD sample-names are different from our/vcf-ismapper expectation

No-special-characters

No dots |  "-"  => change automatically to "_"

Therefore implement, "_" => 0 or more underscores

**Comment by TimHHH** [comment]
Good to see you found the root cause of this issue.

I have tested the concatenation with adjusted sample names in the header and this does indeed fix the issue, the extension of the files does not matter.

Please update the concat command to the following:
```bcftools concat sample.vcf.gz sample.bcf.gz | bcftools sort -o sample.concat.vcf.gz -W```

What's important here is the ```sort```, which ensures all the variants are in the correct order and will therefore not cause downstream issues. The ```-W``` will automatically generate an index file, unless you prefer to do this separately.

**Comment by abhi18av** [comment]
UPDATE:

As of https://github.com/TORCH-Consortium/MAGMA/pull/215/commits/b31614d23b9f475f2ddb3cbda71ab8331116a43b we are able to correctly merge the delly and ismapper produced files.

Reference: https://cloud.seqera.io/orgs/TORCH/workspaces/XBS-Nextflow/watch/XXMWIqaBmMllW 


====

@TimHHH , thank you for the tip! I tried it in the pipeline but the version of `bcftools` we're using does not allow `-W` or even `--write-index` so for the moment, I have tweaked the script to stay within the version boundaries.

File: docs/_quarto.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  title: "MAGMA"
+
+  sidebar:
+      style: "docked"
+      search: true
+      contents:
+        - text: "Usage"
+          href: usage.qmd
+        - text: "Output"
+          href: output.qmd
```
Comment (abhi18av) [inline docs/_quarto.yml:14]: I've added the quarto-based website structure.

File: modules/utils/patch_tbprofiler_sv_output.nf [hunk 2]
```diff
@@ -0,0 +1,21 @@
+process UTILS_PATCH_TBPROFILER_SV_OUTPUT {
```
Comment (abhi18av) [inline modules/utils/patch_tbprofiler_sv_output.nf:1]: This process encapsulates the script created by @Wilma-chan

File: samplesheet/test.3samples.csv [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Sample,R1,R2
```
Comment (abhi18av) [inline samplesheet/test.3samples.csv:1]: Test samples from the original magma publication.

File: bin/bcftools_merge__delly.sh [hunk 4]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
+
+# Create temporary directory
+tmp_dir="interim"
+
+mkdir $tmp_dir
+
+# Extract unique sample prefixes from filenames
+prefixes=( $(for file in *.bcf.gz *.vcf.gz; do basename "$file" | cut -d '.' -f 1,2; done | sort -u) )
```
Comment (abhi18av) [inline bin/bcftools_merge__delly.sh:9]: I've extracted this script, contributed by @vrennie into a separate testable script in the `bin` folder.

File: bin/patch_tbprofiler_sv_output.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+
+import argparse
+import subprocess
+import pandas as pd
+import pysam
+import os
+import glob
+import json
+
```
Comment (abhi18av) [inline bin/patch_tbprofiler_sv_output.py:11]: I have modularized @Wilma-chan script into different functions, easier to test and change.

File: bin/samplesheet_validation.py [hunk 6]
```diff
@@ -1,81 +1,85 @@
...
+
+def validate_fastq_files(r1, r2, line_num):
+    if r1 == r2:
+        print(f'Row {line_num}: {r1}, {r2} - DUPLICATED fastq file specified')
+        return False
+    return True
+
+def set_default_value(field, default_value):
+    return field if field and field.strip() else default_value
+
```
Comment (abhi18av) [inline bin/samplesheet_validation.py:31]: Updated the samplesheet validation as discussed in our last call together.

**Review by abhi18av — COMMENTED** [review]
Hi all,

Here's a quick summary of all the changes done so far in this branch.

The current milestone is that the script has been accommodated completely in the pipeline. Needed to make extensive changes to make it fit as per the flow of files and data.

https://cloud.seqera.io/orgs/TORCH/workspaces/XBS-Nextflow/watch/4BhVp17xDGaH95


image

**Comment by abhi18av** [comment]
@vrennie  and @mdediegofuertes okay so I'm resuming from today.

I see a few changes on the patch script and can you please confirm that the code was working as expected on your side?

**Comment by vrennie** [comment]
@abhi18av just doing some final tests!

**Pull request merged**